File: day2/python/day2.py
import os

# Get the directory where the Python script is located
current_dir = os.path.dirname(os.path.abspath(__file__))

# Construct the relative file path
file_path = os.path.join(current_dir, 'day2.txt')

part_1 = 0

def checkLine(line: list):
    safe = True
    inc = -1
    prevVal = -1
    fail = 0
    fail_ind = -1
    skipAllowed = 1  # Allow skipping one invalid value
    for i, value in enumerate(line):
        value = int(value)
        
        if prevVal == -1:
            prevVal = value
            continue

        # Detect invalid conditions
        isEqualOrTooClose = abs(prevVal - value) < 1  # Adjacent values must differ by at least 1
        isTooFarApart = abs(prevVal - value) > 3     # Adjacent values must differ by at most 3
        isInvalidDecreasing = (prevVal <= value and inc == 0) 
        isInvalidIncreasing = (prevVal >= value and inc == 1)

        # Handle invalid values (with option to skip one)
        if (isEqualOrTooClose or isTooFarApart or isInvalidDecreasing or isInvalidIncreasing):
            safe = False
            fail += 1
            if skipAllowed > 0:
                skipAllowed -= 1  # Skip this invalid value
                fail_ind = i
                continue  # Don't update prevVal, effectively "skipping" this value

        # Update trend direction if undecided
        if inc == -1:
            inc = 0 if prevVal > value else 1

        prevVal = value

    return safe, fail, fail_ind

# Processing the file
part_1 = 0
with open(file_path, 'r') as file:
    for line in file:
        # Handle line and split on space
        cur_line = line.strip().split(" ")

        # Call checkLine for both cases
        checkLine0, lineFail0, fail_ind0 = checkLine(cur_line)
        checkLine1, lineFail1, fail_ind1 = checkLine(cur_line[1::])  # Skip the first element for the second check

        # Determine if the line is valid based on conditions
        if (lineFail0 <= 0) or (lineFail1 <= 0):
            part_1 += 1
        # if checkLine0:
        #     part_1 += 1
        if fail_ind0 != -1:
            cur_line.remove(cur_line[fail_ind0])
            # print(test)
            checkLineRemove, lineFailRemove, fail_indRemove = checkLine(cur_line)
            if checkLineRemove:
                part_1 += 1

print(part_1)

File: day2/cpp/day2.cpp
#include <iostream>
#include <fstream>
#include <fstream>
#include <sstream>
#include <cstdlib>

using namespace std;

int main() {
    ifstream file("day2.txt");
    
    // Check if the file was opened successfully
    if (!file.is_open()) {
        cerr << "Unable to open file" << endl;
        return 1;
    } else {
        cout << "file opened \n";
    }

    // Read the file line by line
    int rows = 0;
    int part1 = 0;
    int part2 = 0;
    char buffer[256];
    while (file.getline(buffer, 256)) {
        rows++;

        istringstream lineStream(buffer);
        int value;
        int prevVal = -1;
        int inc = -1;
        int safe = 1;
        // int safe2 = 1;
        int life = 2;

        // Part 1
        // while (lineStream >> value) {
        //     if (prevVal == -1) {
        //         prevVal = value; // Initialize prevVal
        //         continue;
        //     }
        //     int lostLife = 0;

        //     // Detect invalid sequences or trends
        //     bool isEqualOrTooClose = abs(prevVal - value) < 1; // Adjacent values must differ by at least 1
        //     bool isTooFarApart = abs(prevVal - value) > 3;     // Adjacent values must differ by at most 3
        //     bool isInvalidDecreasing = (prevVal <= value && inc == 0); // Invalid in decreasing trend
        //     bool isInvalidIncreasing = (prevVal >= value && inc == 1); // Invalid in increasing trend

        //     if (isEqualOrTooClose || isTooFarApart || isInvalidDecreasing || isInvalidIncreasing) {
        //         // Commented out for part 2
        //         // safe = -1;
        //         // break;
        //         life --;
        //         if (life == 0) {
        //             safe = -1;
        //             break;
        //         }
        //         lostLife = 1;
        //     }

   

        //     // Update trend if undecided
        //     if (inc == -1) {
        //         inc = (prevVal > value) ? 0 : 1;
        //     }
            
        //     if (lostLife == 0) {
        //         prevVal = value; // Update prevVal for the next iteration
        //     }
        // }

        // if (safe == 1) {
        //     part1 ++;
        // }

        // Gave up using cpp
        while (lineStream >> value) {
            // Handle the first value
            if (prevVal == -1) {
                prevVal = value; // Initialize prevVal with the first value
                continue; // Skip further checks for the first value
            }

            int lostLife = 0;

            // Detect invalid sequences or trends
            bool isEqualOrTooClose = abs(prevVal - value) < 1; // Adjacent values must differ by at least 1
            bool isTooFarApart = abs(prevVal - value) > 3;     // Adjacent values must differ by at most 3
            bool isInvalidDecreasing = (prevVal <= value && inc == 0); // Invalid in decreasing trend
            bool isInvalidIncreasing = (prevVal >= value && inc == 1); // Invalid in increasing trend

            // Handle illegal value
            if (isEqualOrTooClose || isTooFarApart || isInvalidDecreasing || isInvalidIncreasing) {
                life--; // Decrease life for an illegal value
                if (life == 0) {
                    safe = -1; // Mark sequence as unsafe
                    break;     // Exit the loop
                }
                lostLife = 1; // Indicate that a life was lost due to an illegal value
            }

            // Update trend if undecided
            if (inc == -1) {
                inc = (prevVal > value) ? 0 : 1;
            }

            // Only update prevVal if no life was lost
            if (lostLife == 0) {
                prevVal = value; // Update prevVal for the next iteration
            }
        }

        if (safe == 1) {
            part1++; // Increment part1 if the sequence is safe
        }

    }

    cout << part1 << "\n";
    // cout << rows;

}

